### PR TITLE
251 lesson scheduler service

### PR DESCRIPTION
**Pull request by ishootdice**
- Created lesson scheduler service which is able to schedule sending of lessons reminding notifications
- Fixed notifications sending(removed duplicates)

### CODE REVIEW

**Review by bochka123 — DISMISSED** [review]

**Review by bochka123 — DISMISSED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                ImageId = shortInfo.Author.ImageId
+            };
+        }
+    }
+    
+    private async Task SaveNotification(List<NewNotificationDto> newNotifications)
+    {
+        await using var scope = _services.CreateAsyncScope();
+        
+        var context = scope.ServiceProvider.GetRequiredService<EasySpeakCoreContext>();
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: you can pass it as parameter

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        await context.Notifications.AddRangeAsync(notifications);
+
+        await context.SaveChangesAsync();
+    }
+
+    public async Task RemindSubscribers(long lessonId)
+    {
+        await using var scope = _services.CreateAsyncScope();
+        
+        var context = scope.ServiceProvider.GetRequiredService<EasySpeakCoreContext>();
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: the same

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    {
+        var currentTime = DateTime.UtcNow;
+
+        var fromTime = currentTime.AddMinutes(_schedulerOptions.CheckPeriod);
+
+        var toTime = currentTime.AddMinutes(_schedulerOptions.CheckPeriod * 2);
+
+        using var scope = _services.CreateScope();
+        
+        var context = scope.ServiceProvider.GetRequiredService<EasySpeakCoreContext>();
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: DI?

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs (hunk 2 above)
Comment (tatianahutii) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: everywhere

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    public long LessonId { get; set; }
+    public int DelayInMinutes { get; set; }
+}
+
+public class LessonSubscribersNotifyDto
+{
+    public Lesson Lesson { get; set; } = null!;
+    public List<User> Subscribers { get; set; } = new();
+    public User Author { get; set; } = null!;
+}
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: separate files

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        {
+            SendToRabbit(item.Email, item.Notification);
+        }
+    }
+
+    private async Task<List<long>> GetCreatedRemindersList(CancellationToken cancellationToken)
+    {
+        using (var scope = _services.CreateScope())
+        {
+            var context = scope.ServiceProvider.GetRequiredService<EasySpeakCoreContext>();
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: Can you move working with context to separate serviceor lessonsService?

**Review by Anatolii-Ivanov — APPROVED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs (hunk 3 above)
Comment (Anatolii-Ivanov) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/LessonSchedulerService.cs]: It's background service, so it's not possible to get this service via usual DI

**Review by tatianahutii — APPROVED** [review]

**Pull request merged**